apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/GraphQLCompiler.kt

```diff
@@ -1,52 +1,174 @@
 import java.io.File
 
 class GraphQLCompiler {
   fun write(args: Arguments) {
-    val ir = args.ir
-    val customTypeMap = args.customTypeMap.supportedTypeMap(ir.typesUsed)
-    val context = CodeGenerationContext(
-        reservedTypeNames = emptyList(),
-        typeDeclarations = ir.typesUsed,
-        customTypeMap = customTypeMap,
-        operationOutput = args.operationOutput,
-        nullableValueType = args.nullableValueType,
-        ir = ir,
-        useSemanticNaming = args.useSemanticNaming,
-        generateModelBuilder = args.generateModelBuilder,
-        useJavaBeansSemanticNaming = args.useJavaBeansSemanticNaming,
-        suppressRawTypesWarning = args.suppressRawTypesWarning,
-        generateVisitorForPolymorphicDatatypes = args.generateVisitorForPolymorphicDatatypes
+    args.outputDir.deleteRecursively()
+    args.outputDir.mkdirs()
+
+    val roots = Roots(args.rootFolders)
+    val metadata = collectMetadata(args.metadata, args.rootProjectDir)
+
+    val (introspectionSchema, schemaPackageName) = getSchemaInfo(roots, args.rootPackageName, args.schemaFile, metadata)
+
+    val generateKotlinModels = metadata?.generateKotlinModels ?: args.generateKotlinModels
+    val userCustomTypesMap = metadata?.customTypesMap ?: args.customTypeMap
+
+    val packageNameProvider = DefaultPackageNameProvider(
+        roots = roots,
+        rootPackageName = args.rootPackageName
     )
 
-    if (args.generateKotlinModels) {
+    val files = args.graphqlFiles
+    checkDuplicateFiles(roots, files)
+
+    val parseResult = GraphQLDocumentParser(
+        schema = introspectionSchema,
+        packageNameProvider = packageNameProvider
+    ).parse(files)
+
+    val ir = IRBuilder(
+        schema = introspectionSchema,
+        schemaPackageName = schemaPackageName,
+        incomingMetadata = metadata,
+        alwaysGenerateTypesMatching = args.alwaysGenerateTypesMatching,
+        generateMetadata = args.generateMetadata
+    ).build(parseResult)
+
+    val operationOutput = ir.operations.map {
+      OperationDescriptor(
+          name = it.operationName,
+          packageName = it.packageName,
+          filePath = it.filePath,
+          source = QueryDocumentMinifier.minify(it.sourceWithFragments)
+      )
+    }.let {
+      args.operationOutputGenerator.generate(it)
+    }
+
+    check(operationOutput.size == ir.operations.size) {
+      """The number of operation IDs (${operationOutput.size}) should match the number of operations (${ir.operations.size}).
+        |Check that all your IDs are unique.
+      """.trimMargin()
+    }
+
+    if (args.operationOutputFile != null) {
+      args.operationOutputFile.writeText(operationOutput.toJson("  "))
+    }
+
+    val customTypeMap = (introspectionSchema.types.values.filter {
+      it is IntrospectionSchema.Type.Scalar && ScalarType.forName(it.name) == null
+    }.map { it.name } + ScalarType.ID.name)
+        .supportedTypeMap(userCustomTypesMap, generateKotlinModels)
```

Comment:
The code around custom scalars is more complex now, I'll most likely try to refactor in a subsequent pull request. `customTypeMap` here contains all the scalar types from the schema and their matching className either defined by the user or by us (for ID and Any).